File: packages/protocol/connectorv1/capability_test.go
package connectorv1

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// Contract layer (docs/TESTING.md section 2): the tunnel gateway verifies in Go
// the capabilities the control plane mints in TypeScript, so both languages run
// this corpus. A token that only one language produces is a broken deployment,
// not a style difference.

const capabilityFixturesDir = "../fixtures/capability/v1"

type capabilityMintFixture struct {
	Name             string `json:"name"`
	KeyID            string `json:"key_id"`
	CapabilityID     string `json:"capability_id"`
	RouteID          string `json:"route_id"`
	ProjectID        string `json:"project_id"`
	BrowserSessionID string `json:"browser_session_id"`
	IssuedAt         int64  `json:"issued_at"`
	ExpiresAt        int64  `json:"expires_at"`
	Token            string `json:"token"`
	Note             string `json:"note"`
}

type capabilityClaimsFixture struct {
	KeyID            string `json:"key_id"`
	CapabilityID     string `json:"capability_id"`
	RouteID          string `json:"route_id"`
	ProjectID        string `json:"project_id"`
	BrowserSessionID string `json:"browser_session_id"`
	IssuedAt         int64  `json:"issued_at"`
	ExpiresAt        int64  `json:"expires_at"`
}

type capabilityVerifyFixture struct {
	Name   string                   `json:"name"`
	Token  string                   `json:"token"`
	Now    int64                    `json:"now"`
	Expect string                   `json:"expect"`
	Claims *capabilityClaimsFixture `json:"claims"`
	Note   string                   `json:"note"`
}

type capabilityManifest struct {
	Protocol        string                    `json:"protocol"`
	Version         int                       `json:"version"`
	Keys            map[string]string         `json:"keys"`
	VerifierKeyring []string                  `json:"verifier_keyring"`
	Mint            []capabilityMintFixture   `json:"mint"`
	Verify          []capabilityVerifyFixture `json:"verify"`
}

func loadCapabilityManifest(t *testing.T) capabilityManifest {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join(capabilityFixturesDir, "manifest.json"))
	if err != nil {
		t.Fatalf("read capability manifest: %v", err)
	}
	var manifest capabilityManifest
	if err := json.Unmarshal(raw, &manifest); err != nil {
		t.Fatalf("parse capability manifest: %v", err)
	}
	if len(manifest.Mint) == 0 || len(manifest.Verify) == 0 {
		t.Fatal("capability corpus is empty")
	}
	return manifest
}

func capabilityKey(t *testing.T, manifest capabilityManifest, keyID string) []byte {
	t.Helper()
	encoded, ok := manifest.Keys[keyID]
	if !ok {
		t.Fatalf("corpus has no key %q", keyID)
	}
	key, err := base64.StdEncoding.DecodeString(encoded)
	if err != nil {
		t.Fatalf("decode key %q: %v", keyID, err)
	}
	return key
}

func capabilityKeyring(t *testing.T, manifest capabilityManifest) CapabilityKeyring {
	t.Helper()
	keyring := CapabilityKeyring{}
	for _, keyID := range manifest.VerifierKeyring {
		keyring[keyID] = capabilityKey(t, manifest, keyID)
	}
	return keyring
}

func TestCapabilityCorpusMintsGoldenTokens(t *testing.T) {
	manifest := loadCapabilityManifest(t)
	for _, fixture := range manifest.Mint {
		t.Run(fixture.Name, func(t *testing.T) {
			token, err := MintCapability(capabilityKey(t, manifest, fixture.KeyID), CapabilityClaims{
				KeyID:            fixture.KeyID,
				CapabilityID:     fixture.CapabilityID,
				RouteID:          fixture.RouteID,
				ProjectID:        fixture.ProjectID,
				BrowserSessionID: fixture.BrowserSessionID,
				IssuedAt:         fixture.IssuedAt,
				ExpiresAt:        fixture.ExpiresAt,
			})
			if err != nil {
				t.Fatalf("mint: %v", err)
			}
			if token.Reveal() != fixture.Token {
				t.Fatalf("token does not match the corpus:\n  got:  %s\n  want: %s", token.Reveal(), fixture.Token)
			}
			if len(fixture.Token) > MaxCapabilityTokenLength {
				t.Fatalf("token of %d characters exceeds the schema bound", len(fixture.Token))
			}
			// The token must satisfy the schema's session_capability pattern,
			// because that is the field it travels in.
			var violations []SchemaViolation
			validateSessionCapability(fixture.Token, "$", &violations)
			if len(violations) > 0 {
				t.Fatalf("token does not satisfy the session_capability schema: %+v", violations)
			}
		})
	}
}

func TestCapabilityCorpusVerification(t *testing.T) {
	manifest := loadCapabilityManifest(t)
	keyring := capabilityKeyring(t, manifest)
	for _, fixture := range manifest.Verify {
		t.Run(fixture.Name, func(t *testing.T) {
			claims, failure := VerifyCapability(keyring, fixture.Token, fixture.Now)
			if fixture.Expect == "valid" {
				if failure != nil {
					t.Fatalf("corpus requires acceptance, got %v", failure)
				}
				if fixture.Claims == nil {
					t.Fatal("corpus accepts the token but records no claims")
				}
				want := *fixture.Claims
				got := capabilityClaimsFixture{
					KeyID:            claims.KeyID,
					CapabilityID:     claims.CapabilityID,
					RouteID:          claims.RouteID,
					ProjectID:        claims.ProjectID,
					BrowserSessionID: claims.BrowserSessionID,
					IssuedAt:         claims.IssuedAt,
					ExpiresAt:        claims.ExpiresAt,
				}
				if got != want {
					t.Fatalf("claims do not match the corpus:\n  got:  %+v\n  want: %+v", got, want)
				}
				return
			}
			if failure == nil {
				t.Fatalf("corpus requires rejection %q, but the token was accepted", fixture.Expect)
			}
			if string(failure.Rejection) != fixture.Expect {
				t.Fatalf("rejection %q, corpus requires %q", failure.Rejection, fixture.Expect)
			}
		})
	}
}

func TestCapabilityRejectsTokensSignedWithAnotherKey(t *testing.T) {
	manifest := loadCapabilityManifest(t)
	claims := CapabilityClaims{
		KeyID:            "stage0-a",
		CapabilityID:     "cap_x",
		RouteID:          "svc_x",
		ProjectID:        "prj_x",
		BrowserSessionID: "brs_x",
		IssuedAt:         1000,
		ExpiresAt:        2000,
	}
	// The attacker owns a key, and names a key identifier the verifier trusts.
	// Only the MAC decides.
	token, err := MintCapability(capabilityKey(t, manifest, "k"), claims)
	if err != nil {
		t.Fatalf("mint: %v", err)
	}
	keyring := CapabilityKeyring{"stage0-a": capabilityKey(t, manifest, "stage0-a")}
	if _, failure := VerifyCapability(keyring, token.Reveal(), 1500); failure == nil {
		t.Fatal("a capability signed with another key was accepted")
	} else if failure.Rejection != CapabilityRejectionBadSignature {
		t.Fatalf("rejection %q, want bad_signature", failure.Rejection)
	}
}

func TestCapabilityIsRedactedInEveryDefaultRepresentation(t *testing.T) {
	manifest := loadCapabilityManifest(t)
	fixture := manifest.Mint[0]
	token, err := MintCapability(capabilityKey(t, manifest, fixture.KeyID), CapabilityClaims{
		KeyID:            fixture.KeyID,
		CapabilityID:     fixture.CapabilityID,
		RouteID:          fixture.RouteID,
		ProjectID:        fixture.ProjectID,
		BrowserSessionID: fixture.BrowserSessionID,
		IssuedAt:         fixture.IssuedAt,
		ExpiresAt:        fixture.ExpiresAt,
	})
	if err != nil {
		t.Fatalf("mint: %v", err)
	}
	encoded, err := json.Marshal(token)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	for name, representation := range map[string]string{
		"String":     token.String(),
		"json":       string(encoded),
		"%v":         fmt.Sprintf("%v", token),
		"%s":         fmt.Sprintf("%s", token),
		"%q":         fmt.Sprintf("%q", token),
		"%#v":        fmt.Sprintf("%#v", token),
		"containing": fmt.Sprintf("%+v", struct{ Capability SensitiveString }{token}),
	} {
		if strings.Contains(representation, fixture.Token) {
			t.Fatalf("%s leaked the capability: %s", name, representation)
		}
		if !strings.Contains(representation, Redacted) {
			t.Fatalf("%s is not redacted: %s", name, representation)
		}
	}
}

func TestCapabilityRefusesShortSigningKeys(t *testing.T) {
	claims := CapabilityClaims{
		KeyID: "k", CapabilityID: "c", RouteID: "r", ProjectID: "p",
		BrowserSessionID: "b", IssuedAt: 1, ExpiresAt: 2,
	}
	if _, err := MintCapability(make([]byte, MinCapabilitySigningKeyBytes-1), claims); err == nil {
		t.Fatal("a short signing key was accepted")
	}
	// A verifier configured with a short key refuses rather than downgrading.
	short := CapabilityKeyring{"k": make([]byte, MinCapabilitySigningKeyBytes-1)}
	full, err := MintCapability(make([]byte, MinCapabilitySigningKeyBytes), claims)
	if err != nil {
		t.Fatalf("mint: %v", err)
	}
	if _, failure := VerifyCapability(short, full.Reveal(), 1); failure == nil ||
		failure.Rejection != CapabilityRejectionUnknownKey {
		t.Fatalf("a short verifier key must refuse as unknown_key, got %v", failure)
	}
}

func TestCapabilityRefusesUnorderedOrOversizedClaims(t *testing.T) {
	key := make([]byte, MinCapabilitySigningKeyBytes)
	base := CapabilityClaims{
		KeyID: "k", CapabilityID: "c", RouteID: "r", ProjectID: "p",
		BrowserSessionID: "b", IssuedAt: 100, ExpiresAt: 200,
	}

	expired := base
	expired.ExpiresAt = base.IssuedAt
	if _, err := MintCapability(key, expired); err == nil {
		t.Fatal("a capability that expires when it is issued was minted")
	}

	oversized := base
	oversized.RouteID = strings.Repeat("r", MaxCapabilityIdentifierLength+1)
	if _, err := MintCapability(key, oversized); err == nil {
		t.Fatal("an over-long route identifier was minted")
	}

	empty := base
	empty.ProjectID = ""
	if _, err := MintCapability(key, empty); err == nil {
		t.Fatal("an empty project identifier was minted")
	}
}

File: packages/protocol/test/capability.test.ts
/**
 * Contract layer (`docs/TESTING.md` section 2): the control plane mints route
 * capabilities in TypeScript and the tunnel gateway verifies them in Go, so
 * both languages run this corpus. A token that only one language produces is a
 * broken deployment, not a style difference.
 */

import assert from "node:assert/strict";
import { inspect } from "node:util";
import test from "node:test";

import {
  CapabilityError,
  MAX_CAPABILITY_IDENTIFIER_LENGTH,
  MAX_CAPABILITY_TOKEN_LENGTH,
  MIN_CAPABILITY_SIGNING_KEY_BYTES,
  mintCapability,
  verifyCapability,
} from "../src/capability.ts";
import type { CapabilityKeyring } from "../src/capability.ts";
import { loadCapabilityManifest } from "../src/fixtures.ts";
import { REDACTED } from "../src/sensitive.ts";
import { validateSessionCapability } from "../src/generated/connector/v1/validate.ts";

const manifest = loadCapabilityManifest();

function key(keyId: string): Uint8Array {
  const encoded = manifest.keys[keyId];
  assert.ok(encoded !== undefined, `corpus has no key ${keyId}`);
  return new Uint8Array(Buffer.from(encoded, "base64"));
}

function keyring(): CapabilityKeyring {
  return new Map(manifest.verifier_keyring.map((keyId) => [keyId, key(keyId)]));
}

function rejectionOf(error: unknown): string {
  assert.ok(error instanceof CapabilityError, `expected a CapabilityError, got ${inspect(error)}`);
  return error.rejection;
}

test("the corpus mints its golden tokens", () => {
  assert.ok(manifest.mint.length > 0, "the capability corpus is empty");
  for (const fixture of manifest.mint) {
    const token = mintCapability(key(fixture.key_id), {
      keyId: fixture.key_id,
      capabilityId: fixture.capability_id,
      routeId: fixture.route_id,
      projectId: fixture.project_id,
      browserSessionId: fixture.browser_session_id,
      issuedAt: fixture.issued_at,
      expiresAt: fixture.expires_at,
    });
    assert.equal(token.reveal(), fixture.token, `${fixture.name} does not match the corpus`);
    assert.ok(
      fixture.token.length <= MAX_CAPABILITY_TOKEN_LENGTH,
      `${fixture.name} exceeds the schema bound`,
    );
    // The token must satisfy the schema's session_capability pattern, because
    // that is the field it travels in.
    const violations: Parameters<typeof validateSessionCapability>[2] = [];
    validateSessionCapability(fixture.token, "$", violations);
    assert.deepEqual(violations, [], `${fixture.name} is not a valid session_capability`);
  }
});

test("the corpus verification outcomes hold", () => {
  const keys = keyring();
  for (const fixture of manifest.verify) {
    if (fixture.expect === "valid") {
      const claims = verifyCapability(keys, fixture.token, fixture.now);
      assert.ok(fixture.claims !== undefined, `${fixture.name} accepts but records no claims`);
      assert.deepEqual(
        {
          key_id: claims.keyId,
          capability_id: claims.capabilityId,
          route_id: claims.routeId,
          project_id: claims.projectId,
          browser_session_id: claims.browserSessionId,
          issued_at: claims.issuedAt,
          expires_at: claims.expiresAt,
        },
        fixture.claims,
        `${fixture.name} claims do not match the corpus`,
      );
      continue;
    }
    assert.throws(
      () => verifyCapability(keys, fixture.token, fixture.now),
      (error: unknown) => rejectionOf(error) === fixture.expect,
      `${fixture.name} must be refused as ${fixture.expect}`,
    );
  }
});

test("a capability signed with another key is refused", () => {
  // The attacker owns a key, and names a key identifier the verifier trusts.
  // Only the MAC decides.
  const forged = mintCapability(key("k"), {
    keyId: "stage0-a",
    capabilityId: "cap_x",
    routeId: "svc_x",
    projectId: "prj_x",
    browserSessionId: "brs_x",
    issuedAt: 1000,
    expiresAt: 2000,
  });
  const keys: CapabilityKeyring = new Map([["stage0-a", key("stage0-a")]]);
  assert.throws(
    () => verifyCapability(keys, forged.reveal(), 1500),
    (error: unknown) => rejectionOf(error) === "bad_signature",
  );
});

test("a minted capability is redacted in every default representation", () => {
  const fixture = manifest.mint[0];
  assert.ok(fixture !== undefined);
  const token = mintCapability(key(fixture.key_id), {
    keyId: fixture.key_id,
    capabilityId: fixture.capability_id,
    routeId: fixture.route_id,
    projectId: fixture.project_id,
    browserSessionId: fixture.browser_session_id,
    issuedAt: fixture.issued_at,
    expiresAt: fixture.expires_at,
  });
  const representations: Record<string, string> = {
    toString: token.toString(),
    "String()": String(token),
    template: `${token}`,
    "JSON.stringify": JSON.stringify(token),
    "JSON.stringify(containing)": JSON.stringify({ capability: token }),
    inspect: inspect(token),
    "inspect(containing)": inspect({ capability: token }),
  };
  for (const [name, representation] of Object.entries(representations)) {
    assert.ok(!representation.includes(fixture.token), `${name} leaked the capability`);
    assert.ok(representation.includes(REDACTED), `${name} is not redacted`);
  }
  assert.equal(token.reveal(), fixture.token, "reveal() must still produce the wire value");
});

test("short signing keys are refused rather than downgraded", () => {
  const claims = {
    keyId: "k",
    capabilityId: "c",
    routeId: "r",
    projectId: "p",
    browserSessionId: "b",
    issuedAt: 1,
    expiresAt: 2,
  };
  assert.throws(() =>
    mintCapability(new Uint8Array(MIN_CAPABILITY_SIGNING_KEY_BYTES - 1), claims),
  );
  const token = mintCapability(new Uint8Array(MIN_CAPABILITY_SIGNING_KEY_BYTES), claims);
  const short: CapabilityKeyring = new Map([
    ["k", new Uint8Array(MIN_CAPABILITY_SIGNING_KEY_BYTES - 1)],
  ]);
  assert.throws(
    () => verifyCapability(short, token.reveal(), 1),
    (error: unknown) => rejectionOf(error) === "unknown_key",
  );
});

test("claims that break the codec's bounds are refused at minting", () => {
  const signingKey = new Uint8Array(MIN_CAPABILITY_SIGNING_KEY_BYTES);
  const base = {
    keyId: "k",
    capabilityId: "c",
    routeId: "r",
    projectId: "p",
    browserSessionId: "b",
    issuedAt: 100,
    expiresAt: 200,
  };
  assert.throws(() => mintCapability(signingKey, { ...base, expiresAt: base.issuedAt }));
  assert.throws(() =>
    mintCapability(signingKey, {
      ...base,
      routeId: "r".repeat(MAX_CAPABILITY_IDENTIFIER_LENGTH + 1),
    }),
  );
  assert.throws(() => mintCapability(signingKey, { ...base, projectId: "" }));
});
